### PR TITLE
Datagen API review feedback

### LINKED ISSUE #3564: New datagen API

### PR DESCRIPTION
**Pull request by robertbastian**
Fixes #3564

### CODE REVIEW

File: provider/datagen/src/baked_exporter.rs [hunk 1]
```diff
@@ -29,11 +31,17 @@
...
+//! There are two ways to use baked data: you can build custom data providers for use with
+//! [`_unstable` constructors](icu_provider::constructors), or you can use it with the
+//! `compiled_data` Cargo feature and constructors.
+//!
+//! ## Custom `DataProvider`
+//!
+//! This allows you to use baked data in custom data pipelines, such as including some baked
+//! data and lazily loading more data from the network.
 //!
 //! ```
```
Comment (sffc) [inline provider/datagen/src/baked_exporter.rs:43]: Nit: This example should reference `icu4x_baked_demo`, too

 > Markus: "icu4x_baked_demo" is in the first example but I don't see it referenced in the second example. Be more clear about how to get from the output directory of the exporter to the input directory of the runtime provider

File: provider/datagen/src/lib.rs [hunk 2]
```diff
@@ -306,6 +306,13 @@ impl std::fmt::Display for CollationHanDatabase {
 }
 
 /// A language's CLDR coverage level.
+///
+/// Note that in CLDR usage, these levels are understood to be additive, i.e. the "basic"
+/// level includes all languages that are either "basic", "moderate", or "modern". In ICU4X on the
+/// other hand, a language is only part of a single coverage level. This allows generating different
+/// data files for different coverage levels without duplicating data. However, the data itself is
+/// still additive (e.g. for fallback to work correctly), so data for moderate (basic)
+/// languages should only be loaded if modern (modern and moderate) data is already present.
```
Comment (sffc) [inline provider/datagen/src/lib.rs:315]: Nit: I prefer to put positive example up front and contrasting example behind it because people sometimes only read the first sentence

```suggestion
/// In ICU4X, these are disjoint sets: a language belongs to a single coverage level. This allows
/// generating different data files for different coverage levels without duplicating data. This
/// contrasts with CLDR usage, where these levels are understood to be additive (i.e., "basic"
/// implying "basic", "moderate", and "modern"). However, the ICU4X data itself is
/// still additive (e.g. for fallback to work correctly), so data for moderate (basic)
/// languages should only be loaded if modern (modern and moderate) data is already present.
```

File: provider/fs/src/export/serializers/bincode.rs [hunk 3]
```diff
@@ -59,7 +59,7 @@ impl AbstractSerializer for Serializer {
 }
 
 impl Serializer {
-    /// Creates a new serializer for [`bincode`].
+    #[doc(hidden)]
```
Comment (sffc) [inline provider/fs/src/export/serializers/bincode.rs:62]: Issue: I prefer avoiding `#[doc(hidden)]` except for (1) private APIs that need to be public for a legitimate reason like databake and (2) deprecated APIs for semver reasons, except that we should really be preferring `#[deprecated]` for that purpose. I don't think this one fits into either category unless you are intending to deprecate it. Is that your intent, and if so, why? Also, if you don't want people to use `Options`, you should make that deprecated as well.

tl;dr, to avoid additional debate, please revert the doc(hidden) changes here.

File: provider/fs/src/export/serializers/json.rs [hunk 4]
```diff
@@ -86,25 +79,17 @@ impl AbstractSerializer for Serializer {
...
     pub fn new(options: Options) -> Self {
         Self {
             style: options.style,
         }
     }
 
-    /// Convenience function to create a JSON serializer with the
-    /// [`StyleOption::Pretty`] format.
+    /// A JSON serializer that pretty-prints the output.
     pub fn pretty() -> Self {
```
Comment (sffc) [inline provider/fs/src/export/serializers/json.rs:90]: Issue: With all the APIs you're making hidden, it's not clear how people build one of these in non-pretty mode. It will appear in docs that `pretty()` is the only constructor. If someone finds `default()`, its behavior won't be documented. Please either un-hide these APIs or add a new constructor like `Json::compact()` to contrast with `Json::pretty()`

File: provider/fs/src/export/serializers/mod.rs [hunk 5]
```diff
@@ -22,6 +22,7 @@ use icu_provider::datagen::*;
 use icu_provider::prelude::*;
 use std::io;
 
+#[doc(hidden)]
```
Comment (sffc) [inline provider/fs/src/export/serializers/mod.rs:25]: Issue: I don't think this should be `#[doc(hidden)]`. It is a public API that is linked to from other APIs and it is useful in terms of seeing what concrete items implement it. The feedback on this trait was:

> Should explain that this is a trait not intended to implement yourself and you should use one of the three built in serializers

File: provider/fs/src/export/serializers/postcard.rs [hunk 6]
```diff
@@ -64,14 +64,8 @@ impl AbstractSerializer for Serializer {
 }
 
 impl Serializer {
-    /// Creates a new serializer for [`postcard`].
+    #[doc(hidden)]
```
Comment (sffc) [inline provider/fs/src/export/serializers/postcard.rs:67]: Nit: Same comments from the other two serializer modules apply here

File: provider/datagen/README.md [hunk 7]
```diff
@@ -80,7 +80,7 @@ $ icu4x-datagen config.json
```
Comment (sffc) [inline provider/datagen/README.md:80]: Did you fix this one?

> DatagenProvider: ExportableProvider link is broken

**Review by sffc — CHANGES_REQUESTED** [review]

File: provider/datagen/README.md (hunk 7 above)
Comment (robertbastian) [reply provider/datagen/README.md:80]: Couldn't reproduce this

**Review by Manishearth — APPROVED** [review]

File: provider/datagen/README.md (hunk 7 above)
Comment (sffc) [reply provider/datagen/README.md:80]: It's on this page: https://unicode-org.github.io/icu4x/docs/icu_datagen/struct.DatagenProvider.html

The link to ExportableProvider is a 404.

**Review by sffc — APPROVED** [review]

File: provider/datagen/README.md (hunk 7 above)
Comment (robertbastian) [reply provider/datagen/README.md:80]: It doesn't reproduce locally, so shouldn't be an actual issue on crates.io.
